extmod/asyncio/core.py

```diff
@@ -51,9 +70,10 @@ def __next__(self):
 
 
 # Pause task execution for the given time (integer in milliseconds, uPy extension)
-# Use a SingletonGenerator to do it without allocating on the heap
-def sleep_ms(t, sgen=SingletonGenerator()):
-    assert sgen.state is None
+# Try not to allocate a SleepHandler on the heap if possible
+def sleep_ms(t, sgen=SleepHandler()):
+    if sgen.state is not None:  # the static one is busy
+        sgen = SleepHandler()
```

Comment:
Could you give some details on when / how the singleton is already in use?
It seems like it'll be confusing that sometimes sleep will allocate and sometimes it won't.

Reply:
Of course. Simply start two tasks and let both of them sleep at the same time.

I don't think it's confusing, this is entirely internal to `sleep_ms`.

Reply:
I very regularly have a large number of tasks with each of them sleeping for different lengths of time... I haven't noticed any issue. 

What problem / symptoms do you see?

Reply:
I initially found this bug via test 12 in `tests/extmod/asyncio_taskgroup.py`, according to my notes, though I don't remember details.

This test no longer errors out, but adding a `print` statement to the `if` clause does show that the problem is still triggered by the test, and that it does cause the sleep time of *some* other task to be modified and/or ignored.

The non-augmented code depends on the scheduler to iterate the "singleton" object that's yielded by `sleep_ms` immediately, i.e.. before any other task that also happens to be scheduled calls `sleep_ms`. This obviously holds most of the time, but not always.

Reply:
> The original code depends on the scheduler to iterate the "singleton" object that's yielded by `sleep_ms` immediately, i.e.. before any other task which also happens to be scheduled calls `sleep_ms`. This consition seems to hold most of the time, but not always.

Ah, this is quite interesting. I had wondered how the Singleton could ever work to provide multiple different sleep function usages.... I can see how multiple tasks started back to back (like in a new taskgroup setup) would be more likely to trigger this situation of multiple definitions before the task loop makes it back around to the start to process the sleeps.